waspc/src/Wasp/Analyzer/Parser/Token.hs

```diff
@@ -1,38 +1,107 @@
-module Wasp.Analyzer.Parser.Token where
-
-import Wasp.Analyzer.Parser.SourcePosition (SourcePosition, calcNextPosition)
-
-data TokenType
-  = TLParen
-  | TRParen
-  | TLSquare
-  | TRSquare
-  | TLCurly
-  | TRCurly
-  | TComma
-  | TColon
-  | TImport
-  | TFrom
-  | TTrue
-  | TFalse
-  | TString String
-  | TInt Integer
-  | TDouble Double
-  | TLQuote String
-  | TQuoted String
-  | TRQuote String
-  | TIdentifier String
-  | TEOF
-  deriving (Eq, Show)
+{-# LANGUAGE DeriveGeneric #-}
 
+module Wasp.Analyzer.Parser.Token
+  ( Token (..),
+    TokenKind (..),
+    tokenKindIsTrivia,
+    showTokenKind,
+  )
+where
+
+import Control.DeepSeq (NFData)
+import Data.Aeson (ToJSON)
+import GHC.Generics (Generic)
+
+-- | The kind of token
+--
+-- This makes no distinction between value-containing tokens and atomic tokens
+-- (i.e. "String" and "KwImport") to make kind comparison easier. To recover the
+-- value associated with a token, you need the context provided by a "Token".
+data TokenKind
+  = White
+  | -- | Newlines (only \n) are separated from whitespace to accomodate line-break
+    -- based error recovery, if that is ever implemented in the future
+    Newline
+  | Comment
+  | LParen
+  | RParen
+  | LSquare
+  | RSquare
+  | LCurly
+  | RCurly
+  | Comma
+  | Colon
+  | KwImport
+  | KwFrom
+  | KwTrue
+  | KwFalse
+  | String
+  | Int
+  | Double
+  | -- | "{= <identifier>"
+    LQuote
+  | -- | "<identifier> =}"
+    RQuote
+  | Quoted
+  | -- | Non-keyword identifier
+    Identifier
+  | -- | Kind for unexpected characters
+    Error
+  deriving (Eq, Ord, Show, Generic)
+
+instance NFData TokenKind
+
+instance ToJSON TokenKind
+
+-- | A token representing a span of text from the source. For value-containing
+-- tokens, the value can be recovered using @tokenText@.
 data Token = Token
-  { tokenType :: TokenType,
-    tokenStartPosition :: SourcePosition,
-    tokenLexeme :: String
+  { tokenKind :: !TokenKind,
+    -- | The width of the text representing this token. The source position is
+    -- not stored; only the width. Later, offsets into the source is computed
+    -- from an entire tree of tokens, and source position is determined on
+    -- demand (for example, when an offset is going to be displayerd in an error
+    -- message).
+    --
+    -- This is guaranteed to be equivalent to @length . tokenText@, but is
+    -- stored explicitly since it is accessed frequently.
+    tokenWidth :: !Int
```

Comment:
The reason it's stored by width instead of offset was future-proofing if end up implementing incremental re-parsing (only parsing the part of the file that changed on the last edit). Storing only width allows us to re-use all tokens from previous parses except the ones near where the text changed.

Reply:
I am trying to make sense of the sentence in the comment.
It says we store something explicitly since it is accessed frequently. What, the width, right? But it is not like we have tokenText, so we can't calculate it! It sounds from the comment like you are storing `width` additionally next to `tokenText`, but you are not, you are storing only `width`. It is not additional field, it is "the" field needed to figure out what token represents, right?
I mean token now contains nothing but kind and width. So it is not standalone anymore, it needs source + tokens to its left to obtain its text/lexeme.

So coming back to it, I don't understand why say "it is stored explicitly since it is accessed frequently", when it is stored for the reason it is essential to be able to reason about this token. Maybe we should rather explain why we don't store the token's lexeme? That is what interests me in any case! Before Token was consisting of lexeme/value and implicit kind (through data constructor). Now it still has kind but value/lexeme was kicked out and replaced with width. So it has very different role then before, it is not standalone anymore. What is the reasoning behind that?

Reply:
I guess the comments here got out of sync with the code, it shouldn't be talking about `tokenText`. The token doesn't store the lexeme since the CST parser doesn't look at the lexemes. I guess this moves the question to "why doesn't the CST store lexemes?" Storing the lexeme in the CST would make it easier to get lexemes for a syntax node, getting rid of the need to keep passing the source string around. There are two ways I see we could go about this:

- We store a lexeme for every syntax node, which could lead towards somewhat significant memory usage, since the lexeme for each token is duplicated by all of its ancestors.
- We store a lexeme only for syntax nodes with no children, which makes it kind of complicated to get the lexeme for a non-leaf node.

Anyways, I guess the reason for tokens not containing lexemes anymore is just that they don't have to, because accessing the lexeme behind a token is more of an edge case (for error reporting or for finding the value for an AST literal node) than being needed commonly.

For now, I've removed mention of `tokenText` in these comments. I haven't added this explanation in yet though, I'll wait to hear your thoughts on it.

Reply:
Ok interesting!
Could you make a NOTE block next to `data Token` explaining what you wrote above?
I am a bit tempted to removed this dependency on `source` and have Token (or SyntaxNode) contain this information, but I am sure points you made above are valid.
I don't have time to deeper evaluate the options, nor do we have time to make any changes, and since this works ok, let's leave it as it is, but let's write this NOTE that will explain reasoning behind and mention possible other ways to do it and pros/cons, as you did above, so in the future we can pick it up from here if we wish!